README.md
# bsdznc

[![Cookbook Version](https://img.shields.io/cookbook/v/bsdznc.svg)](https://supermarket.chef.io/cookbooks/bsdznc)
[![Apache 2.0 License](https://img.shields.io/github/license/rxrc/bsdznc.svg)](./LICENSE.txt)
[![Dependency Status](https://img.shields.io/gemnasium/rxrc/bsdznc.svg)](https://gemnasium.com/rxrc/bsdznc)
[![Build Status](https://img.shields.io/travis/rxrc/bsdznc.svg)](https://travis-ci.org/rxrc/bsdznc)
[![Coverage Status](https://img.shields.io/codecov/c/github/rxrc/bsdznc.svg)](https://codecov.io/github/rxrc/bsdznc)

## Description

FreeBSD ZNC server.

### Usage

Include either the `bsdznc::default` recipe,
or use the LWRP directly, e.g.,

```ruby
bsdznc 'default' do
  action :create
end

# Now make sure the SSL certificate exists, then start ZNC.

bsdznc 'default' do
  action :start
end
```

You must provide a valid SSL certificate.
If you use the default recipe, this is handled by the [certificate cookbook].
You must have an encrypted data bag under `certificates` with id `znc`.
See the [certificate cookbook] docs for how to create this.

Access the web admin at `https://example.com/`.
The ZNC server will listen for IRC connections on port `194`.

A minimal `znc.conf` will be created
with user `admin` and password `password`.
This should be changed immediately.

The `znc.conf` file is only created if it does not exist:
it will not be touched otherwise.
You should backup this file regularly as it contains
all server and user settings configured through the web admin.

[certificate cookbook]: https://supermarket.chef.io/cookbooks/certificate


## Requirements

### Platforms

* Freebsd (~> 10.1)

### Cookbooks:

* certificate (~> 1.0.0)
* sysctl (~> 0.6.2)

## Attributes

Attribute | Description | Default | Choices
----------|-------------|---------|--------
`node['bsdznc']['user']` | `User to run znc as.` | "znc" |
`node['bsdznc']['group']` | `Group to run znc as.` | "znc" |
`node['bsdznc']['uid']` | `User id for the znc user.` | 194 |
`node['bsdznc']['gid']` | `Group id for the znc group.` | 194 |
`node['bsdznc']['version']` | `ZNC version to use for the default config.` | "1.6.0" |
`node['bsdznc']['config_path']` | `Path to the ZNC configuration directory.` | "/usr/local/etc/znc" |
`node['bsdznc']['cert_file']` | `SSL certificate filename.` | "znc.pem" |
`node['bsdznc']['irc_port']` | `Port to listen on for IRC connections.` | 194 |
`node['bsdznc']['web_port']` | `Port to listen on for https connections.` | 443 |
`node['bsdznc']['uri_prefix']` | `Prefix to serve the web admin under.` | "/" |

## Recipes

* bsdznc::default - Install, configure, and start the ZNC server.

## Resources

* [bsdznc](#bsdznc)

### bsdznc



#### Actions

- create: create the znc server. Default action.
- destroy: stop and destroy the znc server.
- nothing: do nothing.
- restart: restart the znc server.
- start: start the znc server.
- stop: stop the znc server.

## Development and Testing

### Source Code

The [bsdznc source](https://github.com/rxrc/bsdznc)
is hosted on GitHub.
To clone the project run

```bash
$ git clone https://github.com/rxrc/bsdznc.git
```

### Rake

Run `rake -T` to see all Rake tasks.

```
rake all                          # Run all tasks
rake doc                          # Build documentation
rake foodcritic                   # Lint Chef cookbooks
rake kitchen:all                  # Run all test instances
rake kitchen:default-freebsd-101  # Run default-freebsd-101 test instance
rake readme                       # Generate README.md from _README.md.erb
rake rubocop                      # Run RuboCop
rake rubocop:auto_correct         # Auto-correct RuboCop offenses
rake spec                         # Run RSpec code examples
rake test                         # Run kitchen integration tests
rake yard                         # Generate YARD Documentation
```

### Thor

Run `thor -T` to see all Thor tasks.

### Guard

Guard tasks have been separated into the following groups:

- `doc`
- `lint`
- `unit`
- `integration`

By default, Guard will generate documentation, lint, and run unit tests.
The integration group must be selected manually with `guard -g integration`.

## Contributing

Please submit and comment on bug reports and feature requests.

To submit a patch:

1. Fork it (https://github.com/rxrc/bsdznc/fork).
2. Create your feature branch (`git checkout -b my-new-feature`).
3. Make changes. Write and run tests.
4. Commit your changes (`git commit -am 'Add some feature'`).
5. Push to the branch (`git push origin my-new-feature`).
6. Create a new Pull Request.

## License

Copyright 2015 Evan Sosenko

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

## Warranty

This software is provided "as is" and without any express or
implied warranties, including, without limitation, the implied
warranties of merchantibility and fitness for a particular
purpose.


CHANGELOG.md
# ChangeLog

## 0.0.1

Initial release.


spec/znc_spec.rb
require 'spec_helper'

describe 'bsdznc::znc' do
  let(:chef_run) { ChefSpec::ServerRunner.new.converge(described_recipe) }

  it 'creates the znc server' do
    expect(chef_run).to configure_znc('default')
  end

  it 'creates the ssl certificate' do
    expect(chef_run).to create_certificate_manage('znc')
  end

  it 'starts the znc server' do
    expect(chef_run).to start_znc('default')
  end
end


spec/default_spec.rb
require 'spec_helper'

describe 'bsdznc::default' do
  cached(:chef_run) { ChefSpec::ServerRunner.new.converge(described_recipe) }

  it 'installs the znc server' do
    expect(chef_run).to install_znc('default')
  end

  it 'creates the ssl certificate' do
    expect(chef_run).to create_certificate_manage('znc')
  end

  it 'starts the znc server' do
    expect(chef_run).to start_znc('default')
  end
end


spec/providers/default_spec.rb
require 'spec_helper'

describe 'bsdznc_test::create' do
  let(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['bsdznc'])
      .converge(described_recipe)
  end

  it 'installs znc' do
    expect(chef_run).to install_package('znc')
  end

  it 'creates the znc user and group' do
    expect(chef_run).to create_user('znc').with(system: true)
    expect(chef_run).to create_group('znc').with(system: true)
  end

  it 'creates the configs directory' do
    expect(chef_run).to create_directory('/usr/local/etc/znc/configs')
  end

  context 'when no existing configuration' do
    before(:each) do
      allow(::File).to receive(:exist?).and_call_original
      allow(::File).to receive(:exist?)
        .with('/usr/local/etc/znc/configs/znc.conf').and_return(true)
    end

    it 'does not overwrite the current configuration' do
      expect(chef_run).to_not create_template(
        '/usr/local/etc/znc/configs/znc.conf')
    end
  end

  context 'when no existing configuration' do
    before(:each) do
      allow(::File).to receive(:exist?).and_call_original
      allow(::File).to receive(:exist?)
        .with('/usr/local/etc/znc/configs/znc.conf').and_return(false)
    end

    it 'creates the default configuration' do
      expect(chef_run).to create_template(
        '/usr/local/etc/znc/configs/znc.conf')
    end
  end

  it 'enables and loads the portacl module' do
    expect(chef_run).to create_directory('/etc/rc.conf.d')
    expect(chef_run).to create_file('/etc/rc.conf.d/mac_portacl.conf')
    expect(chef_run).to run_execute('kldload -n mac_portacl.ko')
  end

  it 'allows znc to bind to reserved ports' do
    expect(chef_run).to(
      apply_sysctl_param('net.inet.ip.portrange.reservedhigh')
      .with(value: 0))
    expect(chef_run).to(
      apply_sysctl_param('security.mac.portacl.enabled')
      .with(value: 1))
    expect(chef_run).to(
      apply_sysctl_param('security.mac.portacl.suser_exempt')
      .with(value: 1))
    expect(chef_run).to(
      apply_sysctl_param('security.mac.portacl.rules')
      .with(value: 'uid:194:tcp:194,uid:194:tcp:443'))
  end

  it 'enables znc' do
    expect(chef_run).to enable_service('znc')
  end
end

describe 'bsdznc_test::start' do
  cached(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['bsdznc'])
      .converge(described_recipe)
  end

  it 'starts znc' do
    expect(chef_run).to start_service('znc')
  end
end

describe 'bsdznc_test::stop' do
  cached(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['bsdznc'])
      .converge(described_recipe)
  end

  it 'stops znc' do
    expect(chef_run).to stop_service('znc')
  end
end

describe 'bsdznc_test::restart' do
  cached(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['bsdznc'])
      .converge(described_recipe)
  end

  it 'stops and starts znc' do
    expect(chef_run).to stop_service('znc')
    expect(chef_run).to start_service('znc')
  end
end

describe 'bsdznc_test::destroy' do
  cached(:chef_run) do
    ChefSpec::ServerRunner.new(step_into: ['bsdznc'])
      .converge(described_recipe)
  end

  it 'stops and disables znc' do
    expect(chef_run).to disable_service('znc')
    expect(chef_run).to stop_service('znc')
  end

  it 'returns the portacl rules to default' do
    expect(chef_run).to(
      apply_sysctl_param('security.mac.portacl.enabled')
      .with(value: 0))
    expect(chef_run).to(
      apply_sysctl_param('net.inet.ip.portrange.reservedhigh')
      .with(value: 1023))
    expect(chef_run).to remove_sysctl_param('security.mac.portacl.enabled')
  end

  it 'disables the portacl module' do
    expect(chef_run).to delete_file('/etc/rc.conf.d/mac_portacl.conf')
  end

  it 'removes the znc user and group' do
    expect(chef_run).to remove_user('znc')
    expect(chef_run).to remove_group('znc')
  end

  it 'deletes the znc configuration directory' do
    expect(chef_run).to delete_directory('/usr/local/etc/znc')
  end

  it 'removes znc' do
    expect(chef_run).to remove_package('znc')
  end
end


recipes/znc.rb
#
# Cookbook Name:: bsdznc
# Recipe:: znc
#
# Copyright 2015 Evan Sosenko
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

bsdznc 'default' do
  action :create
  subscribes :restart, 'certificate_manage[znc]'
end

certificate_manage 'znc' do
  cert_path node['bsdznc']['cert_path']
  cert_file node['bsdznc']['cert_file']
  combined_file true
  owner node['bsdznc']['user']
  group node['bsdznc']['group']
end

bsdznc 'default' do
  action :start
end


providers/default.rb
#
# Cookbook Name:: bsdznc
# Provider:: default
#
# Copyright 2015 Evan Sosenko
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

def whyrun_supported?
  true
end

use_inline_resources

action :create do
  run_context.include_recipe 'sysctl'

  package 'znc' do
    action :install
  end

  user node['bsdznc']['user'] do
    uid node['bsdznc']['uid']
    home '/nonexistent'
    shell '/usr/sbin/nologin'
    system true
    action :create
  end

  group node['bsdznc']['group'] do
    gid node['bsdznc']['gid']
    system true
    action :create
  end

  directory "#{node['bsdznc']['config_path']}/configs" do
    user node['bsdznc']['user']
    group node['bsdznc']['group']
    mode '0750'
    recursive true
    action :create
  end

  template "#{node['bsdznc']['config_path']}/configs/znc.conf" do
    source 'znc.conf.erb'
    user node['bsdznc']['user']
    group node['bsdznc']['group']
    mode '0640'
    action :create
    not_if do
      ::File.exist?("#{node['bsdznc']['config_path']}/configs/znc.conf")
    end
  end

  directory '/etc/rc.conf.d' do
    user 'root'
    group 'wheel'
    mode '0755'
    action :create
  end

  file '/etc/rc.conf.d/mac_portacl.conf' do
    content 'mac_portacl_load="YES"'
    user 'root'
    group 'wheel'
    mode '0644'
    action :create
  end

  execute 'kldload -n mac_portacl.ko' do
    user 'root'
  end

  sysctl_param 'net.inet.ip.portrange.reservedhigh' do
    value 0
    action :apply
  end

  sysctl_param 'security.mac.portacl.enabled' do
    value 1
    action :apply
  end

  sysctl_param 'security.mac.portacl.suser_exempt' do
    value 1
    action :apply
  end

  sysctl_param 'security.mac.portacl.rules' do
    value ["uid:#{node['bsdznc']['uid']}:tcp:#{node['bsdznc']['irc_port']}",
           "uid:#{node['bsdznc']['uid']}:tcp:#{node['bsdznc']['web_port']}"
          ].join(',')
    action :apply
  end

  service 'znc' do
    action :enable
  end
end

action :start do
  service 'znc' do
    action :start
  end
end

action :stop do
  service 'znc' do
    action :stop
  end
end

action :restart do
  service 'znc' do
    action [:stop, :start]
  end
end

action :destroy do
  service 'znc' do
    action [:stop, :disable]
  end

  sysctl_param 'security.mac.portacl.enabled' do
    value 0
    action [:apply, :remove]
  end

  sysctl_param 'net.inet.ip.portrange.reservedhigh' do
    value 1023
    action :apply
  end

  file '/etc/rc.conf.d/mac_portacl.conf' do
    action :delete
  end

  directory node['bsdznc']['config_path'] do
    action :delete
  end

  user node['bsdznc']['user'] do
    action :remove
  end

  group node['bsdznc']['group'] do
    action :remove
  end

  package 'znc' do
    action :remove
  end
end


metadata.rb
name             'bsdznc'
maintainer       'Evan Sosenko'
maintainer_email 'razorx@evansosenko.com'
license          'Apache 2.0'
description      'FreeBSD ZNC server.'
source_url       'https://github.com/rxrc/bsdznc'
issues_url       'https://github.com/rxrc/bsdznc/issues'
version          '0.0.1'

supports 'freebsd', '~> 10.1'

depends 'certificate', '~> 1.0.0'
depends 'sysctl', '~> 0.6.2'

recipe 'bsdznc::default', 'Install, configure, and start the ZNC server.'

provides 'service[bsdznc]'

attribute 'bsdznc/user',
          display_name: 'ZNC user',
          description: 'User to run znc as.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: 'znc'

attribute 'bsdznc/group',
          display_name: 'ZNC group',
          description: 'Group to run znc as.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: 'znc'

attribute 'bsdznc/uid',
          display_name: 'ZNC user id',
          description: 'User id for the znc user.',
          type: 'numeric',
          recipes: ['cafe-core::default'],
          default: 194

attribute 'bsdznc/gid',
          display_name: 'ZNC group id',
          description: 'Group id for the znc group.',
          type: 'numeric',
          recipes: ['cafe-core::default'],
          default: 194

attribute 'bsdznc/version',
          display_name: 'ZNC version',
          description: 'ZNC version to use for the default config.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: '1.6.0'

attribute 'bsdznc/config_path',
          display_name: 'ZNC config path',
          description: 'Path to the ZNC configuration directory.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: '/usr/local/etc/znc'

attribute 'bsdznc/cert_file',
          display_name: 'SSL certificate',
          description: 'SSL certificate filename.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: 'znc.pem'

attribute 'bsdznc/irc_port',
          display_name: 'IRC port',
          description: 'Port to listen on for IRC connections.',
          type: 'numeric',
          recipes: ['cafe-core::default'],
          default: 194

attribute 'bsdznc/web_port',
          display_name: 'Web admin port',
          description: 'Port to listen on for https connections.',
          type: 'numeric',
          recipes: ['cafe-core::default'],
          default: 443

attribute 'bsdznc/uri_prefix',
          display_name: 'URI prefix',
          description: 'Prefix to serve the web admin under.',
          type: 'string',
          recipes: ['cafe-core::default'],
          default: '/'
